### PR TITLE
IKEA Trådfri binding

### PR DESCRIPTION
**Pull request by dsundberg**
Initial features:
- Discovery of Gateway
- Discovery of bulbs
- Control On/Off
- Brightness 
- Change between 3 color temperatures for White Spectrum bulbs

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Time for me to go shopping at IKEA ;-)

File: addons/binding/org.openhab.binding.ikeatradfri/ESH-INF/thing/thing-types.xml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    <!-- Sample Thing Type -->
+    <bridge-type id="gateway">
+        <label>IKEA Tradfri Gateway Binding</label>
+        <description>Binding for the IKEA Tradfri Gateway</description>
+
+        <config-description>
+            <parameter name="host" type="text" required="true">
+                <label>Host</label>
+                <description>Location for the weather information.
```
Comment (hakan42) [inline addons/binding/org.openhab.binding.ikeatradfri/ESH-INF/thing/thing-types.xml]: Seems to be a copy&paste error on the parameter description 😄

Comment (dsundberg) [inline addons/binding/org.openhab.binding.ikeatradfri/ESH-INF/thing/thing-types.xml]: =) a lot of copy paste yes, I'll put it on the todo... thanks!

**Comment by dsundberg** [comment]
Color temperature is now dimmable instead of 3 fixed settings.

**Comment by luxerama** [comment]
@dsundberg hey, thanks so much for starting this binding. Much appreciated!
Two questions:
1. I am moving the jar into the addons folder and i can see openhab2 loding it, however the bridge does not appear in the paper ui.
```
10:57:20.119 [DEBUG] [org.openhab.binding.ikeatradfri     ] - ServiceEvent REGISTERED - {o
rg.eclipse.smarthome.io.transport.mdns.discovery.MDNSDiscoveryParticipant}={component.name
=org.openhab.binding.ikeatradfri.discovery.gateway, component.id=198, service.id=331, serv
ice.bundleid=207, service.scope=bundle} - org.openhab.binding.ikeatradfri
10:57:20.138 [DEBUG] [org.openhab.binding.ikeatradfri     ] - BundleEvent STARTED - org.op
enhab.binding.ikeatradfri
10:57:20.143 [INFO ] [org.apache.felix.fileinstall        ] - Started bundle: file:/usr/sh
are/openhab2/addons/org.openhab.binding.ikeatradfri-2.1.0-SNAPSHOT.jar
```
2. Will we be able to steer other ZLL rgb bulbs with it or will it be limited to the different whites?

If there is anything i can do to help please let me know.

Cheers.

**Comment by luxerama** [comment]
Re point no 1, for anyone else having the same issue, reset the bridge and try again.

**Comment by mavris** [comment]
Hi guys,
A quick question. Is the RC working with Wifi? Or I will need extra equipment to integrate it with OH?

**Comment by kaikreuzer** [comment]
The RC is using ZLL as well, no wifi. If the bridge can be convinced to send out events from RC commands, it should be possible to include them directly into openHAB as well, which would be really cool (one could e.g. use the dimmers for volume control or other stuff...)

**Comment by kaikreuzer** [comment]
Hi @dsundberg! I started looking into the code and instead of adding review comments, I have created https://github.com/dsundberg/openhab2-addons/pull/1 with a few changes - please let me know, if you have any questions on them. Please also check if this PR might break any functionality as I didn't test the changes with real devices yet.

**Comment by sjka** [comment]
First of all: Thank you very much for starting this binding! Really cool to have this.

I didn't have a look at the code yet, but there are two thing I noticed immediately when trying it out:

* there is a README.md missing (which you know anyway, I'm pretty sure) - that makes it harder for people wanting to try it out because you have to look up all the configuration parameters yourself. At least when configuring it in the DSL files.
* The bulb IDs matter. Usually the user is free to pick names for Thing's IDs and the technical identifier for the device is a configuration property. 

Apart from that the first impression from a users perspective actually was quite nice: I just worked 😄  (@kaikreuzer: I am running it with dsundberg#1 included, so it doesn't look like it broke it). So thanks again for taking the time to write this and contribute it here!

**Comment by kaikreuzer** [comment]
Thanks for checking @sjka, glad to hear that I didn't break anything. 
I just hope that @dsundberg will return soon as he has fallen silent since more than 3 weeks by now...

**Comment by sjka** [comment]
Just have a look at his profile picture - seems like he's got twins...
So @dsundberg, you got my deepest respect that you managed to free enough time for writing this binding at all!

**Comment by kaikreuzer** [comment]
Let's have @dsundberg enjoy his twins then - as I have learned that I can push commits to this PR, I just took the liberty to do so. So everybody is invited to review and test this code!

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+            client.setEndpoint(endPoint);
+            CoapHandler handler = new CoapHandler() {
+                @Override
+                public void onLoad(CoapResponse response) {
+                    if (response.isSuccess()) {
+                        logger.debug("COAP PUT Successful to: {}", url);
+                        future.complete(response.getResponseText());
+                    } else {
+                        logger.debug("COAP PUT Error: {} for {}", response.getCode().toString(), url);
+                        future.completeExceptionally(null);
```
Comment (sjka) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: this throws an NPE! Add a RuntimeException as in the GET method above.

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+                    }
+                    removeAsyncClient(client);
+                }
+
+                @Override
+                public void onError() {
+                    logger.debug("COAP PUT Error");
+                    future.completeExceptionally(null);
```
Comment (sjka) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: NPE!

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+                    removeAsyncClient(client);
+                }
+            };
+            addAsyncClient(client);
+            client.put(handler, payload, MediaTypeRegistry.TEXT_PLAIN);
+            return future;
+        } catch (URISyntaxException e) {
+            logger.warn("COAP URI exception: {}", e.getMessage());
+            future.completeExceptionally(null);
```
Comment (sjka) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: NPE!

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+                pendingObserve.add(childThing.getUID());
+            }
+        }
+    }
+
+    @Override
+    public void childHandlerDisposed(ThingHandler childHandler, Thing childThing) {
+        logger.debug("Child handler disposed: {}", childThing.getThingTypeUID().toString());
+        deviceDiscoverHelper(childThing.getUID().getId());
+    }
```
Comment (sjka) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: shouldn't it also `stopObserve(...)` here?

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+            CoapClient client = new CoapClient(uri);
+            client.setEndpoint(endPoint);
+            CoapHandler handler = new CoapHandler() {
+                @Override
+                public void onLoad(CoapResponse response) {
+                    logger.debug("COAP Observe: \noptions: {}\npayload: {} ", response.getOptions().toString(),
+                            response.getResponseText());
+                    if (response.isSuccess()) {
+                        try {
+                            listener.onDataUpdate(parser.parse(response.getResponseText()));
```
Comment (sjka) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: Funny observation: When I change the valued from the Trådfri App, then this here gets notified on the _first_ change. Afterwards it looses the contact.

I have no clue about the californium API, but from wildly guessing adding this line appears to fix it:

```
observeRelationMap.get(thingUID).reregister();
```

So it looks like such an observation relation needs to be re-registered once it fired. But as I said, I actually have no clue...

Comment (sjka) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: Hm. Just learned that this _shouldn't_ be necessary. Sorry bout the spam 😞

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java (hunk 2 above)
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: fixed

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java (hunk 3 above)
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: fixed

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java (hunk 4 above)
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: fixed

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java (hunk 5 above)
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: I think so - and have added it.
Why is a discovery triggered here immediately, do you know?

Comment (sjka) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriGatewayHandler.java]: I can't make any sense of it. My vote goes to dropping it here.

**Comment by sjka** [comment]
There is one thing I just noticed at home: The binding stopped working completely. It doesn't get any updates anymore, and when trying to send commands I end up with timeouts like this:

```
java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1975)[:1.8.0_112]
	at org.openhab.binding.ikeatradfri.handler.IkeaTradfriGatewayHandler$2.onError(IkeaTradfriGatewayHandler.java:192)[201:org.openhab.binding.ikeatradfri:2.1.0.201705191643]
	at org.eclipse.californium.core.CoapClient$MessageObserverImpl.failed(CoapClient.java:999)[201:org.openhab.binding.ikeatradfri:2.1.0.201705191643]
	at org.eclipse.californium.core.CoapClient$MessageObserverImpl.onTimeout(CoapClient.java:960)[201:org.openhab.binding.ikeatradfri:2.1.0.201705191643]
	at org.eclipse.californium.core.coap.Message.setTimedOut(Message.java:512)[201:org.openhab.binding.ikeatradfri:2.1.0.201705191643]
	at org.eclipse.californium.core.coap.Request.setTimedOut(Request.java:485)[201:org.openhab.binding.ikeatradfri:2.1.0.201705191643]
	at org.eclipse.californium.core.network.stack.ReliabilityLayer$RetransmissionTask.run(ReliabilityLayer.java:368)[201:org.openhab.binding.ikeatradfri:2.1.0.201705191643]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_112]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_112]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)[:1.8.0_112]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)[:1.8.0_112]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_112]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_112]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_112]
```

Of course, this is the NPE that you already fixed. Nevertheless, this is strange as the iOS app and my second instance from within the IDE are still working. So I'm wondering whether we are missing a mechanism to detect and recover "broken" connections somehow....?

File: addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriBulbHandler.java [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        }
+
+        LightProperties setBrightness(PercentType b) {
+            settings.add(TRADFRI_DIMMER, new JsonPrimitive(Math.round(b.floatValue() / 100.0f * 254)));
+            return this;
+        }
+
+        PercentType getBrightness() {
+            int b = settings.get(TRADFRI_DIMMER).getAsInt();
+            if (b == 1) {
```
Comment (sjka) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriBulbHandler.java]: Does this block make sense to anybody here?

Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriBulbHandler.java]: Depends on preference I guess.
b can vary between 0..254 and I guess the author wants 0% only represented by `b=0`. `b=1` is a very low brightness but obviously not 0%, hence the manual 1%.

Comment (sjka) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriBulbHandler.java]: ah, got it.
thanks!

Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.ikeatradfri/src/main/java/org/openhab/binding/ikeatradfri/handler/IkeaTradfriBulbHandler.java]: @sjka see edit

**Comment by marschallphilipp** [comment]
Hi @kaikreuzer, is the binding to be merged into master tree soon? I´m excited :-)

**Comment by kaikreuzer** [comment]
Closing it as it is superseded by https://github.com/eclipse/smarthome/pull/3653.

See [here](https://community.openhab.org/t/ikea-tradfri-gateway/26135/61?u=kai) for details.

**Pull request closed without merging**